File: 2.add-two-numbers.js
/*
 * @lc app=leetcode id=2 lang=javascript
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
function addTwoNumbers(l1, l2, c = null) {
  let val = 0;

  let l1next = null;
  if (l1) {
    val += l1.val;
    l1next = l1.next;
  }

  let l2next = null;
  if (l2) {
    val += l2.val;
    l2next = l2.next;
  }

  if (c && c.val >= 10) {
    val += 1;
    c.val -= 10;
  }

  const l = new ListNode(val);

  if (c) {
    c.next = l;
  }

  if (l1next || l2next || l.val >= 10) {
    addTwoNumbers(l1next, l2next, l);
  }

  return l;
}
// @lc code=end

File: 2.add-two-numbers.rs
/*
 * @lc app=leetcode id=2 lang=rust
 *
 * [2] Add Two Numbers
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn add_two_numbers(
        l1: Option<Box<ListNode>>,
        l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        Self::add_list_nodes(l1, l2, 0)
    }

    fn add_list_nodes(
        l1: Option<Box<ListNode>>,
        l2: Option<Box<ListNode>>,
        carry: i32,
    ) -> Option<Box<ListNode>> {
        match (l1, l2, carry) {
            (None, None, 0) => None,
            (None, None, c) => Some(Box::new(ListNode::new(c))),
            (Some(l), None, _) | (None, Some(l), _) => {
                let sum = l.val + carry;
                Some(Box::new(ListNode {
                    next: Self::add_list_nodes(l.next, None, sum / 10),
                    val: sum % 10,
                }))
            }
            (Some(l1), Some(l2), _) => {
                let sum = l1.val + l2.val + carry;
                Some(Box::new(ListNode {
                    next: Self::add_list_nodes(l1.next, l2.next, sum / 10),
                    val: sum % 10,
                }))
            }
        }
    }
}
// @lc code=end
